src/persist-client/src/impl/state.rs

```diff
@@ -201,6 +197,26 @@ where
         }
         self.since = since;
     }
+
+    fn push_batch(&mut self, keys: &[String], desc: &Description<T>) {
+        // Sneaky optimization! If there are no keys in this batch, we can
+        // simply extend the description of the last one (assuming there is
+        // one). This will be an absurdly common case in actual usage because
+        // empty batches are used to communicate progress and most things will
+        // be low-traffic.
+        if keys.is_empty() {
+            if let Some((_, last_desc)) = self.trace.last_mut() {
```

Comment:
Didn't you mention that it only extends the upper of a batch if it doesn't have keys, meaning we should check whether what we get from `last_mut()` here is actually empty?